File: LeetCode/3484/golang/lc3484.go
package lc3484

import (
	"strconv"
	"strings"
)

// Runtime complexity: O(1) for each query.
// Auxiliary space: O(|SetCell|).
// Subjective level: medium.
type Spreadsheet struct {
	Cells map[string]int
}

func Constructor(rows int) Spreadsheet {
	s := new(Spreadsheet)
	s.Cells = make(map[string]int)
	return *s
}

func (this *Spreadsheet) SetCell(cell string, value int) {
	this.Cells[cell] = value
}

func (this *Spreadsheet) ResetCell(cell string) {
	this.SetCell(cell, 0)
}

func (this *Spreadsheet) GetValue(formula string) int {
	parts := strings.Split(formula[1:], "+")
	return this.evaluate(parts[0]) + this.evaluate(parts[1])
}

func (this *Spreadsheet) evaluate(text string) int {
	if converted, err := strconv.Atoi(text); err == nil {
		return converted
	}
	value, isOk := this.Cells[text]
	if !isOk {
		return 0
	}
	return value
}

File: LeetCode/3484/python/lc3484.py
from typing import List
from collections import defaultdict

# Runtime complexity: amortized O(1) (for each query).
# Auxiliary space complexity: O(1).
# Subjective level: medium
class Spreadsheet:
    def __init__(self, rows: int) -> None:
        self.data = defaultdict(int)

    def setCell(self, cell: str, value: int) -> None:
        self.data[cell] = value

    def resetCell(self, cell: str) -> None:
        self.setCell(cell, 0)

    def getValue(self, formula: str) -> int:
        """Evaluates a formula in the form "=X+Y", where both X and Y are either
        cell references OR non-negative integers. Returns the computed sum."""
        lhs, rhs = formula[1:].split("+")
        return self._evaluate(lhs) + self._evaluate(rhs)

    def _evaluate(self, cell: str) -> int:
        if cell.isdigit():
            return int(cell)
        return self.data[cell]
